### PR TITLE
Enforce Event bus restrictions at register time.

### PR DESCRIPTION
**Pull request by rhyshort**
## What
Enforce event bus restrictions when a subscriber is being registered to the bus.

## Why 
Enforcing the event bus restrictions when registering means the `IllegalAccessException` thrown
when attempting to invoke a method on the subscriber should not occur during normal operation, so it becomes safe to throw a `RuntimeException` when an `IllegalAccessException` is caught.

## How
Make sure subscribing instances conform to the following restrictions when the register method is
called or throw an `IllegalArgumentException`.

- Subscriber class *must* be public
- Methods with the @Subscribe annotation *must* be public.
- Method with the @Subscribe annotation *must* have only one parameter.

Note: Previously the method and subscriber only needed to be visible to the EventBus class, this
has now been restricted to enforce public access for the subscription method and subscription class. This makes it easier to perform validation if

## Testing

New tests added to the EventBusTest which test exception cases.

### CODE REVIEW

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 1]
```diff
@@ -88,9 +105,16 @@ public void register(Object object) {
         do {
             for (Method m : listenerClass.getDeclaredMethods()) {
                 if (m.isAnnotationPresent(com.cloudant.sync.event.Subscribe.class)) {
+
+                    if (!Modifier.isPublic(m.getModifiers())){
+                        throw new IllegalArgumentException("Subscriber method with @Subscribe is required to be public");
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Exception message should name the method which is at fault (fully qualified)

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 2]
```diff
@@ -88,9 +105,16 @@ public void register(Object object) {
...
+                    }
+
                     Class[] params = m.getParameterTypes();
                     if (params.length == 1) {
                         methods.add(new SubscriberMethod(m, params[0]));
+                    } else {
+                        throw new IllegalArgumentException("Subscriber method is required to have only 1 parameter");
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: As above

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 3]
```diff
@@ -19,13 +19,18 @@
...
 import org.junit.Test;
+import org.junit.rules.ExpectedException;
 
 import java.util.ArrayList;
 import java.util.List;
 
 public class EventBusTest {
 
+    @Rule
+    public ExpectedException thrown= ExpectedException.none();
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: spacing around `=`

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 4]
```diff
@@ -74,12 +75,28 @@ public void post(Object event) {
...
      *
+     * <p>
+     *     The following restrictions are placed on the subscriber instance.
+     *     <ul>
+     *         <li>The class of the subscriber <strong>must</strong> be public</li>
+     *         <li>Methods annotated with <code>@Subscribe</code> <strong>must</strong> be public</li>
+     *         <li>Methods annotated with <code>@Subscribe</code> <strong>must</strong> have exactly <strong>one</strong> parameter</li>
+     *     </ul>
+     *
+     * </p>
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: This end tag is what is breaking the javadoc build.

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 5]
```diff
@@ -162,19 +167,178 @@ public void multipleEvents() throws Exception {
...
+
+    @Test
+    public  void privateSubscriber() throws Exception {
+        thrown.expect(IllegalArgumentException.class);
+        thrown.expectMessage("Event Subscriber class needs to be public");
+
+        bus.register(new PrivateSubscriber());
+
+        // Post an event
+        TestEvent e = new TestEvent();
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: None of these error cases need to post events

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 6]
```diff
@@ -162,19 +167,178 @@ public void multipleEvents() throws Exception {
...
+
+        @Subscribe
+        public void throwOnEvent(TestEvent event) throws Exception {
+            throw new Exception("Test subscriber exception");
+        }
+    }
+
+
+    @Test
+    public  void privateSubscriber() throws Exception {
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: A comment per test like the existing ones in this package would be helpful

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 7]
```diff
@@ -162,19 +167,178 @@ public void multipleEvents() throws Exception {
...
+
+    }
+
+    protected static class ProtectedSubScriber{
+    }
+
+    static class PackageProtectedSubscriber {
+    }
+
+    public static class PakageProtectedMethodSubscriber {
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: Pakage → Package

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 2 above)
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Should we consider doing additional checks here that the parameter is one of the notification classes. I believe they are all in the same package now so we could check that maybe?

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 8]
```diff
@@ -162,19 +167,178 @@ public void multipleEvents() throws Exception {
...
+        thrown.expectMessage("Subscriber method is required to have only 1 parameter");
+
+        bus.register(new SubscriberMethod0Params());
+
+        // Post an event
+        TestEvent e = new TestEvent();
+        bus.post(e);
+    }
+
+    public static class SubscriberMethod2Params{
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: Can we put all of these statics in one place? There are some existing ones at the top and one in the middle

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java [hunk 9]
```diff
@@ -162,19 +167,178 @@ public void multipleEvents() throws Exception {
...
+
+
+    @Test
+    public  void privateSubscriber() throws Exception {
+        thrown.expect(IllegalArgumentException.class);
+        thrown.expectMessage("Event Subscriber class needs to be public");
+
+        bus.register(new PrivateSubscriber());
+
+        // Post an event
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: Remove, this isn't necessary when testing the exception.

Comment (ricellis) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: And for the other exception test cases.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 10]
```diff
@@ -74,12 +75,28 @@ public void post(Object event) {
...
      * @param object the instance that will be notified when an event is posted
+     * @throws IllegalArgumentException if the subscribing class does not adhere to the restrictions listed.
+     *
      */
     public void register(Object object) {
 
         Class<?> listenerClass = object.getClass();
 
+        if(!Modifier.isPublic(listenerClass.getModifiers())){
+            throw new IllegalArgumentException("Event Subscriber class needs to be public");
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Exception message should name the class which is at fault (fully qualified)

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 2 above)
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Possibly this leads back to the marker interface I suggested when we made these classes for 1.0. Then we can just do an assignable check.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 10 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Done in af4d3d8

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 1 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Done in af4d3d8

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 2 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Done in af4d3d8

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 4 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: fixed in ffc25b7

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java (hunk 5 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: removed in 98551a3

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java (hunk 3 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: fixed in 7910d5d

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java (hunk 9 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: removed in 98551a3

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java (hunk 8 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: moved to end of file in 20d42da

File: cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java (hunk 7 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/test/java/com/cloudant/sync/event/EventBusTest.java]: fixed in 3290b76

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 11]
```diff
@@ -74,12 +76,29 @@ public void post(Object event) {
      * {@link com.cloudant.sync.event.Subscribe}.
      * </p>
      *
+     * <p>
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Small nitpick but the tags aren't indented properly which makes this hard to follow

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 12]
```diff
@@ -88,9 +107,21 @@ public void register(Object object) {
         do {
             for (Method m : listenerClass.getDeclaredMethods()) {
                 if (m.isAnnotationPresent(com.cloudant.sync.event.Subscribe.class)) {
+
+                    if (!Modifier.isPublic(m.getModifiers())){
+                        throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                "Subscriber method %s#%s is required to be public",
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Can the class be anonymous here as well?

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 13]
```diff
@@ -88,9 +107,21 @@ public void register(Object object) {
...
+                                listenerClass.getCanonicalName(), m.getName()));
+                    }
+
                     Class[] params = m.getParameterTypes();
                     if (params.length == 1) {
                         methods.add(new SubscriberMethod(m, params[0]));
+                    } else {
+                        throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                                "Subscriber method  %s#%s is required to have only 1 parameter",
```
Comment (tomblench) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: As above

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 12 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: No they can't. Anonymous will always be rejected by the first check because they cannot be made public.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 11 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Fixed in c0df327

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+ * and limitations under the License.
+ *
+ */
+
+package com.cloudant.sync.event.notifications;
+
+/**
+ * Marker interface for events.
+ */
+public interface Notification {
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java]: is it weird to have `Notification`s instead of `Event`s on an `EventBus`? In either case the doc should be updated to say what type the parameter of a `@Subscribe` annotated method should be.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java [hunk 15]
```diff
@@ -74,12 +78,29 @@ public void post(Object event) {
...
+     *
      */
     public void register(Object object) {
 
         Class<?> listenerClass = object.getClass();
 
+        if(!Modifier.isPublic(listenerClass.getModifiers())){
+            throw new IllegalArgumentException(String.format(Locale.ENGLISH,
+                    "Event Subscriber class %s needs to be public",
+                    listenerClass.isAnonymousClass() ? "Anonymous" : listenerClass.getCanonicalName()));
```
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: Why not just use `listenerClass.getName()` for these? It would avoid the need to do the anonymous check because an anonymous class will be returned with a `$n` reference which will make it easier for people to identify the offending class.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java (hunk 14 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java]: I thought it made more sense to be `Notification` rather than event, because its in the `notifications` package.

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java (hunk 15 above)
Comment (rhyshort) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/EventBus.java]: done in 18c3fa7

File: cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java (hunk 14 above)
Comment (ricellis) [inline cloudant-sync-datastore-core/src/main/java/com/cloudant/sync/event/notifications/Notification.java]: Well I feel like an `EventBus` should have `Event`s, but there are many other `Event` classes so it is probably less confusing to have a more unique name. Given that the package is notifications, probably best to stick with `Notification`.

**Comment by ricellis** [comment]
@rhyshort I think the 1 remaining change on this is to add the `Notification` type to the `event.md` documentation.

**Comment by rhyshort** [comment]
@ricellis done in 50f4317
